pkg/config/store/contextch.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"context"
+	"sync"
+)
+
+// contextCh is a utility type of outgoing channel with context. It can send messages
+// as long as the context is valid.
+type contextCh struct {
```

Comment:
I'm a bit confused about why we need this type: you can just close a channel, I don't think we need a context to handle closing it too.

Reply:
No, closing channels won't help. Sending on closed channel causes a runtime panic (https://golang.org/ref/spec#Send_statements), that's not wanted here. Here want to skip sending to that channel (ignore it) in that case.

Reply:
I really think the ownership semantics are wrong here, and the watching for a channel closure is happening in the wrong spot. The worker needs to be signaled to stop listening, but instead you're trying to have the sender figure out when to stop pushing data.

----------

At a minimum use the `Done() <-chan struct{}` in the context during construction, not the context itself: you don't need or use it, and it's an odd pattern. Having two channels, one to receive data and one to signal finished, is totally normal in go. What's odd is having a channel and a context together that you use to accomplish the same thing.

```go
type (
    receiver struct {
        events chan<- BackendEvent
        done   <-chan Struct{}
    }

    receivers struct {
        m  sync.RWMutex
        ls []*listener 
    }
)

func (rs *receivers) Send(e BackendEvent) {
	rs.m.RLock()
 	for _, r := range rs.list {
 		select {
 		case <-r.done:
 		case r.events <- ev:
 		}
 	}
	rs.mu.RUnlock()
}
```

However, even then I'm not sure this is correct: if you have a select over two channels and both are valid (i.e. both the events channel is open to receive an event and the context is done) the channel used in the select is picked pseudo randomly, so it's possible that you still attempt the send and panic as a result.